feat: Add --custom-jinja-template argument to pass a custom chat template for SGLang

KrishnanPrash opened the pull request:
#### Overview:

This PR adds support for custom chat templates in Dynamo, allowing users to override a model's default chat template with a custom Jinja template file. This feature enables fine-grained control over prompt formatting without modifying the model repository.

Example Workflow:
```
#Assuming nats & etcd are up and running.
> python -m dynamo.frontend --http-port 8000 &
> python -m dynamo.sglang --model Qwen/Qwen2.5-0.5B-Instruct \
    --skip-tokenizer-init --trust-remote-code \
    --custom-jinja-template ~/custom_chat_template.jinja 
```

Most of this PR leverages the etcd/NATS/MDC groundwork introduced in #2829. 

To test this feature, two tests have been added:
1. ~~`test_custom_jinja_template_mdc()` (CLI -> MDC path): Starts a model with the custom jinja template and pulls the MDC (from etcd) and template (from NATS) and verifies the contents.~~ Removed to limit blast radius of etcd/NATS removal. Any replacement test will be added as a part of a future PR.
2. `test_custom_template_verification()` (MDC -> Inference path): Loads a mocker model + custom `generate()` function. The `generate()` function validates whether the template has been applied by looking for a prefix (`CUSTOM_TEMPLATE_ACTIVE|`) in the received request prompt.

Edit: Adding comprehensive testing.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added an optional CLI flag to specify a custom Jinja template path for prompt formatting.
  * The runtime now propagates and uses this custom template when registering language models, enabling project-specific prompt templates without code changes.
  * Default behavior remains unchanged if the flag is not provided, ensuring backward compatibility and easy experimentation with template variations.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

KrishnanPrash committed:
Adding flag to SGLang + passing it thru register_llm

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

KrishnanPrash committed:
Merge branch 'main' into kprashanth/custom_template_sglang

KrishnanPrash commented on components/backends/sglang/src/dynamo/sglang/args.py:
```diff
@@ -191,6 +198,7 @@ def parse_args(args: list[str]) -> Config:
         migration_limit=parsed_args.migration_limit,
         tool_call_parser=tool_call_parser,
         reasoning_parser=reasoning_parser,
+        custom_jinja_template=parsed_args.custom_jinja_template,
     )
```
Correct me if I am wrong, but the implementation for `--custom-jinja-template` for vLLM does not do any validation before passing it to `register_llm()`. 

Both workflows should benefit from the following early validation check in lib/bindings/python/rust/lib.rs:
```rust
    let custom_template_path_owned = custom_template_path
        .map(|s| {
            let path = PathBuf::from(s);
            if !path.exists() {
                return Err(PyErr::new::<pyo3::exceptions::PyFileNotFoundError, _>(
                    format!("Custom template file does not exist: {}", path.display()),
                ));
            }
            Ok(path)
        })
        .transpose()?;
```

KrishnanPrash replied:
Added support for file path expansions when the file path is provided as a string.

rmccorm4 committed:
Merge branch 'main' into kprashanth/custom_template_sglang

KrishnanPrash committed:
Added Path Expansion Support + Unit Tests (CLI->MDC + MDC->Template)

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

KrishnanPrash committed:
Merge branch 'main' into kprashanth/custom_template_sglang

Signed-off-by: KrishnanPrash <140860868+KrishnanPrash@users.noreply.github.com>

KrishnanPrash committed:
Added Copyrights

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

KrishnanPrash committed:
Pre-Commit Formatting

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

KrishnanPrash committed:
formatting: Cleaning up comments, log instead of print, and adding docstrings

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

KrishnanPrash committed:
Remove additional environment variables

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

KrishnanPrash committed:
Utilized run_serve_deployment(expected_response=...) instead of manually sending a request

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

KrishnanPrash committed:
Removed extra print statement

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

KrishnanPrash committed:
Added comments to requirements to clarify use/need

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

rmccorm4 commented on tests/serve/common.py:
```diff
@@ -75,3 +79,32 @@ def params_with_model_mark(configs: Mapping[str, EngineConfig]):
     return params
+
+
+def get_mdc_from_etcd(mdc_key: str) -> Dict[str, Any]:
```
FYI @grahamking plan is to remove these as soon as alternative discovery APIs are available, but want some interim sanity check in the meantime

KrishnanPrash replied:
Will add a TODO at the top to keep track of this. Thanks for the callout/reminder.

grahamking replied:
Thanks @rmccorm4 ! Will add a top level review.

KrishnanPrash replied:
Also @rmccorm4 , just wanted a clarification on this:
> plan is to remove these as soon as alternative discovery APIs 

When the `ServiceDiscovery` changes get merged, are we going to replace these functions with new helpers that will help us look at the internals of `ServiceDiscovery` or removing this test all together?

grahamking reviewed (changes requested):
Please do not talk to etcd and nats directly from Python. We are working very hard to reduce the blast surface when those two will change.

Instead, validate it the way the client would validate it, using those APIs.

KrishnanPrash committed:
Removed all references to etcd/NATS + Removed test case that pulls MDC

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

grahamking commented:
Thanks for the etcd/nats change. Are you still able to test it?

KrishnanPrash commented:
cc: @grahamking
> Thanks for the etcd/nats change. Are you still able to test it?

Currently, don't know if there is any direct way to test the CLI argument to MDC flow without talking directly to etcd/MDC (or `ServiceDiscovery` in the future). I am definitely open to any ideas you, @GuanLuo, or @rmccorm4 may have.

KrishnanPrash committed:
Adding docs for custom-jinja-template

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

ishandhanani reviewed (changes requested):
From offline convo

```
Can you add a conditional/check for that in args.py? Basically ensure that you cannot ever do --custom-jinja-template with --use-sglang-tokenizer
```

KrishnanPrash committed:
Add guard to prevent --custom-jinja-template and --use-sglang-tokenizer from being used at the same time

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

ishandhanani commented on components/backends/sglang/src/dynamo/sglang/args.py:
```diff
@@ -198,13 +205,29 @@ def parse_args(args: list[str]) -> Config:
+        logging.error(
+            "Cannot use --custom-jinja-template and --use-sglang-tokenizer together. "
+            "--custom-jinja-template requires Dynamo's preprocessor to apply the template, "
+            "while --use-sglang-tokenizer bypasses Dynamo's preprocessor entirely."
```
If you want to use the sglang tokenizer and add a chat template, please use the --chat-template argument from sglang

ishandhanani reviewed (approved):
Please add that small addition to make the error a bit more descriptive. Otherwise lgtm

KrishnanPrash commented:
Added flag to SGLang `README.md` and added a check to prevent `--use-tokenizer-sglang` and `--custom-jinja-template` cannot be used at the same time. Here is what the error message looks like:
```
(venv) root@4845aa7-lcedt:/workspace# python -m dynamo.sglang --model model_name --use-sglang-tokenizer --custom-jinja-template ~/user_template.jinja
...
2025-09-30T22:26:20.329893Z  INFO main.worker: Signal handlers will trigger a graceful shutdown of the runtime
2025-09-30T22:26:20.331523Z ERROR args.parse_args: Cannot use --custom-jinja-template and --use-sglang-tokenizer together. --custom-jinja-template requires Dynamo's preprocessor to apply the template, while --use-sglang-tokenizer bypasses Dynamo's preprocessor entirely.
```

cc: @ishandhanani

KrishnanPrash committed:
more descriptive error logging

Signed-off-by: Krishnan Prashanth <kprashanth@nvidia.com>

KrishnanPrash committed:
Merge branch 'main' into kprashanth/custom_template_sglang
